align/primitive/default/mos.py

```diff
@@ -104,57 +127,74 @@ def single_centered_via(rect):
+            self.addWire( self.pl, None, None, self.gatesPerUnitCell*x+self.gateDummy*self.shared_diff+i,   (y,0), (y,1))
 
+        if self.shared_diff == 1 and (x == 0 or x == x_cells-1):
+            dummy_gates = self.gatesPerUnitCell*x_cells+self.gateDummy if x == x_cells-1 else 0
+            for i in range(self.gateDummy):
+                self.addWire( self.pl, None, None, dummy_gates+i,   (y,0), (y,1))
         # Source, Drain, Gate Connections
 
-        grid_y0 = y*self.m2PerUnitCell + self.finDummy//2-1
-        grid_y1 = grid_y0+(self.finsPerUnitCell - 2*self.finDummy + 2)//2
-        gate_x = x * self.gatesPerUnitCell + self.gatesPerUnitCell // 2
+        #grid_y0 = y*self.m2PerUnitCell + self.finDummy//2-1
+        grid_y0 = y*self.m2PerUnitCell + 1
+        grid_y1 = (y+1)*self.m2PerUnitCell-5
+        #grid_y1 = grid_y0+(self.finsPerUnitCell - 2*self.finDummy + 2)//2-1
+        gate_x = self.gateDummy*self.shared_diff + x * self.gatesPerUnitCell + self.gatesPerUnitCell // 2
         # Connect Gate (gate_x)
-        self.addWire( self.m1, None, None, gate_x , (grid_y0, -1), (grid_y1, 1))
-        self.addVia( self.va, f'{fullname}:G', None, gate_x, (y*self.m2PerUnitCell//2, 1))
+        self.addWire( self.m1, None, None, gate_x , (grid_y1+2, -1), (grid_y1+4, 1))
+        self.addWire( self.pc, None, None, grid_y1+1, (x,1), (x+1,-1))
+        #self.addVia( self.va, f'{fullname}:G', None, gate_x, (y*self.m2PerUnitCell//2, 1))
+        self.addVia( self.va, f'{fullname}:G', None, gate_x, grid_y1+2)
         self._xpins[name]['G'].append(gate_x)
         # Connect Source & Drain
         if reflect:
             _connect_diffusion(gate_x + 1, 'S') #S
-            _connect_diffusion(gate_x - 1, 'D') #D
+            _connect_diffusion(gate_x - 1, 'S') #S
+            _connect_diffusion(gate_x, 'D') #D 
         else:
             _connect_diffusion(gate_x - 1, 'S') #S
-            _connect_diffusion(gate_x + 1, 'D') #D
+            _connect_diffusion(gate_x + 1, 'S') #S 
+            _connect_diffusion(gate_x, 'D') #D
 
     def _connectDevicePins(self, y, connections):
         center_track = y * self.m2PerUnitCell + self.m2PerUnitCell // 2 # Used for m1 extension
+        grid_y1 = (y+1)*self.m2PerUnitCell-5
+        track1 = 0
         for track, (net, conn) in enumerate(connections.items(), start=1):
             contacts = {track for inst, pins in self._xpins.items()
                               for pin, m1tracks in pins.items()
                               for track in m1tracks if (inst, pin) in conn}
             for j in range(self.minvias):
-                current_track = y * self.m2PerUnitCell + len(connections) * j + track
+                if net.startswith('G'):
+                    current_track = grid_y1 + 2 + track1
+                    track1 = track1+1
+                else:
+                    current_track = y * self.m2PerUnitCell + len(connections) * j + track
```

Comment:
build-align reports multiple short errors involving Gate. You might want to review this code using the failures as counterexamples.